### PR TITLE
[feg][aaa][radius] Add support for Radius Dynamic Authorization Extensions (RFC5176) to AAA radius server

### PR DESCRIPTION
**Pull request by emakeev**
Signed-off-by: Evgeniy Makeev <evgeniym@fb.com>

## Summary

Added support for Radius Dynamic Authorization Extensions to builtin AAA Server radius. To enable builtin radius 'with_builtin_radius' is required. The default build creates legacy aaa server which relies on an external radius server (see magma/feg/radius).

## Test Plan

FeG: make precommit;

build & deploy on test OpenWRT AP:
GOOS=linux GOARCH=arm go build -tags link_local_service,with_builtin_radius -o ~/bin/arm/ magma/feg/gateway/services/aaa/aaa_server
scp ~/bin/arm/aaa_server root@192.168.1.1:/sbin/

1) Start server and attach/detach user:

Mon Jul 27 18:44:45 2020 daemon.info aaa_server[14760]: I0728 01:44:45.297077   14760 in_memory.go:147] setting timeout of 300.000000 seconds for session: 3CC5B1AEE81A823D
Mon Jul 27 18:44:45 2020 daemon.info aaa_server[14760]: I0728 01:44:45.306230   14760 auth.go:161] successfully authenticated user: 0001010000000119@wlan.mnc001.mcc001.3gppnetwork.org
Mon Jul 27 18:44:45 2020 daemon.notice hostapd: wlan0: CTRL-EVENT-EAP-SUCCESS2 2e:73:09:75:e2:0d
Mon Jul 27 18:44:45 2020 daemon.info hostapd: wlan0: STA 2e:73:09:75:e2:0d WPA: pairwise key handshake completed (RSN)
Mon Jul 27 18:44:45 2020 daemon.notice hostapd: wlan0: AP-STA-CONNECTED 2e:73:09:75:e2:0d
Mon Jul 27 18:44:45 2020 daemon.info hostapd: wlan0: STA 2e:73:09:75:e2:0d RADIUS: starting accounting session 3CC5B1AEE81A823D
Mon Jul 27 18:44:45 2020 daemon.info hostapd: wlan0: STA 2e:73:09:75:e2:0d IEEE 802.1X: authenticated - EAP type: 23 (unknown)
Mon Jul 27 18:44:45 2020 daemon.info aaa_server[14760]: I0728 01:44:45.326227   14760 accounting.go:81] setting timeout of 300.000000 seconds for session: 3CC5B1AEE81A823D
Mon Jul 27 18:44:45 2020 daemon.info aaa_server[14760]: I0728 01:44:45.326638   14760 acct.go:143] successfully handled Start Accounting Request for session_id:"3CC5B1AEE81A823D" apn:"C6-41-1E-21-EF-55:OpenWrt_MDO" mac_addr:"2E-73-09-75-E2-0D"
...
Mon Jul 27 18:45:01 2020 daemon.notice hostapd: wlan0: AP-STA-DISCONNECTED 2e:73:09:75:e2:0d
Mon Jul 27 18:45:01 2020 daemon.info hostapd: wlan0: STA 2e:73:09:75:e2:0d RADIUS: stopped accounting session 3CC5B1AEE81A823D
Mon Jul 27 18:45:01 2020 daemon.info aaa_server[14760]: I0728 01:45:01.727215   14760 acct.go:163] Interim Update: octets_in:9104 octets_out:7057 packets_in:85 packets_out:45 ctx:<session_id:"3CC5B1AEE81A823D" apn:"C6-41-1E-21-EF-55:OpenWrt_MDO" mac_addr:"2E-73-09-75-E2-0D" >
Mon Jul 27 18:45:01 2020 daemon.info aaa_server[14760]: I0728 01:45:01.938095   14760 acct.go:143] successfully handled Stop Accounting Request for session_id:"3CC5B1AEE81A823D" apn:"C6-41-1E-21-EF-55:OpenWrt_MDO" mac_addr:"2E-73-09-75-E2-0D"

2) Set session timeout to 30s, attach user & wait 4 session timeout to trigger DAE disconnect:

Mon Jul 27 18:54:45 2020 daemon.info hostapd: wlan0: STA 2e:73:09:75:e2:0d IEEE 802.1X: authenticated - EAP type: 23 (unknown)
Mon Jul 27 18:54:45 2020 daemon.info aaa_server[15013]: I0728 01:54:45.385586   15013 accounting.go:81] setting timeout of 30.000000 seconds for session: 94867A6D4C610393
Mon Jul 27 18:54:45 2020 daemon.info aaa_server[15013]: I0728 01:54:45.385847   15013 acct.go:143] successfully handled Start Accounting Request for session_id:"94867A6D4C610393" apn:"C6-41-1E-21-EF-55:OpenWrt_MDO" mac_addr:"2E-73-09-75-E2-0D"
Mon Jul 27 18:55:15 2020 daemon.info aaa_server[15013]: I0728 01:55:15.393003   15013 accounting.go:224] AAA session timeout for SID: 94867A6D4C610393, IMSI: 001010000000119
Mon Jul 27 18:55:15 2020 daemon.notice hostapd: wlan0: AP-STA-DISCONNECTED 2e:73:09:75:e2:0d
Mon Jul 27 18:55:15 2020 daemon.info hostapd: wlan0: STA 2e:73:09:75:e2:0d RADIUS: stopped accounting session 94867A6D4C610393
Mon Jul 27 18:55:15 2020 daemon.info aaa_server[15013]: I0728 01:55:15.607805   15013 acct.go:163] Interim Update: octets_in:8136 octets_out:6253 packets_in:65 packets_out:42 ctx:<session_id:"94867A6D4C610393" apn:"C6-41-1E-21-EF-55:OpenWrt_MDO" mac_addr:"2E-73-09-75-E2-0D" >
Mon Jul 27 18:55:15 2020 daemon.warn aaa_server[15013]: W0728 01:55:15.608638   15013 accounting.go:116] Accounting Stop: Session 94867A6D4C610393 is not found
Mon Jul 27 18:55:15 2020 daemon.info aaa_server[15013]: I0728 01:55:15.608904   15013 acct.go:143] successfully handled Stop Accounting Request for session_id:"94867A6D4C610393" apn:"C6-41-1E-21-EF-55:OpenWrt_MDO" mac_addr:"2E-73-09-75-E2-0D"
Mon Jul 27 18:55:15 2020 daemon.info aaa_server[15013]: I0728 01:55:15.609923   15013 dae_builtin.go:60] DAE Disconnect Response Code Disconnect-ACK

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

File: feg/gateway/services/aaa/radius/dae/dae_builtin.go [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+	"layeh.com/radius"
+	"layeh.com/radius/rfc2865"
+	"layeh.com/radius/rfc2866"
+
+	"magma/feg/cloud/go/protos/mconfig"
+	"magma/feg/gateway/services/aaa/protos"
+	aaa_radius "magma/feg/gateway/services/aaa/radius"
+)
+
+type daeServerCfg struct {
```
Comment (uri200) [inline feg/gateway/services/aaa/radius/dae/dae_builtin.go:30]: I may be missing something but I am just curious why did you add to the name the suffix `cfg` instead of using `deaServerBuiltin`

File: feg/gateway/services/aaa/radius/dae/dae_builtin.go [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+// NewDAEServicer returns servicer for built in DAE service
+func NewDAEServicer(cfg *mconfig.RadiusConfig) DAE {
+	return daeServerCfg{aaa_radius.ValidateConfigs(cfg)}
+}
+
+// Disconnect is DAE's Disconnect Messages equivalent
+func (s daeServerCfg) Disconnect(aaaCtx *protos.Context) error {
+	if s.RadiusConfig == nil {
+		glog.V(1).Info("nil DAE configuration")
```
Comment (uri200) [inline feg/gateway/services/aaa/radius/dae/dae_builtin.go:42]: when are we supposed to use V=1 in magma/feg?

File: feg/gateway/services/aaa/radius/dae/dae_builtin.go [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+	rr := radius.Request{
+		Packet: &radius.Packet{
+			Code:       radius.CodeDisconnectRequest,
+			Identifier: 0,
+			Secret:     s.RadiusConfig.GetSecret(),
+		},
+	}
+	rr.Add(rfc2866.AcctSessionID_Type, radius.Attribute(aaaCtx.GetSessionId()))
+	rr.Add(rfc2865.CallingStationID_Type, radius.Attribute(aaaCtx.GetMacAddr()))
+	// TBD, should it be Exchange or just
```
Comment (uri200) [inline feg/gateway/services/aaa/radius/dae/dae_builtin.go]: can you provide background of what is this `TBD` note about?

File: feg/gateway/services/aaa/radius/dae/dae_external.go [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+	"magma/feg/cloud/go/protos/mconfig"
+	"magma/feg/gateway/registry"
+	"magma/feg/gateway/services/aaa/protos"
+)
+
+type extDAEServer struct{}
+
+// NewDAEServicer returns default servicer using external, registry based service
+func NewDAEServicer(*mconfig.RadiusConfig) DAE {
```
Comment (uri200) [inline feg/gateway/services/aaa/radius/dae/dae_external.go:28]: why do you need RadiusConfig if you are not using it on the creation

File: feg/gateway/services/aaa/radius/dae/dae_external.go [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+	return extDAEServer{}
+}
+
+// Disconnect is DAE's Disconnect Messages equivalent
+func (extDAEServer) Disconnect(aaaCtx *protos.Context) error {
+	conn, err := registry.GetConnection(registry.RADIUS)
+	if err == nil {
```
Comment (uri200) [inline feg/gateway/services/aaa/radius/dae/dae_external.go:35]: I got confused here for a while. Why did you reverse the condition here?. 
```
if err != nil {
	return err
}
...
return nil

**Review by uri200 — COMMENTED** [review]
I checked a document from Cisco about the DAE and I understood is used to change some attributes from the user once this authenticate. But why are we just implementing the `Disconnect` method only>

File: feg/gateway/services/aaa/radius/dae/dae_builtin.go (hunk 2 above)
Comment (emakeev) [reply feg/gateway/services/aaa/radius/dae/dae_builtin.go:42]: lightweight debug logging

File: feg/gateway/services/aaa/radius/dae/dae_builtin.go (hunk 3 above)
Comment (emakeev) [reply feg/gateway/services/aaa/radius/dae/dae_builtin.go]: Exchange is (send & rcv) * retries, I was wondering if send & forget would be enough here since we cannot really do anything on error anyway, but the note itself is a leftover, will remove.

File: feg/gateway/services/aaa/radius/dae/dae_builtin.go (hunk 1 above)
Comment (emakeev) [reply feg/gateway/services/aaa/radius/dae/dae_builtin.go:30]: to emphasize, it's just a config & has no state

File: feg/gateway/services/aaa/radius/dae/dae_external.go (hunk 4 above)
Comment (emakeev) [reply feg/gateway/services/aaa/radius/dae/dae_external.go:28]: to keep API identical between built in & external

File: feg/gateway/services/aaa/radius/dae/dae_external.go (hunk 5 above)
Comment (emakeev) [reply feg/gateway/services/aaa/radius/dae/dae_external.go:35]: less lines, less returns (1 vs 2), I think, it's subjectively easier to read. If it's really looks confusing - LMK, I'll change it.

**Review by emakeev — COMMENTED** [review]
> I checked a document from Cisco about the DAE and I understood is used to change some attributes from the user once this authenticate. But why are we just implementing the `Disconnect` method only>

Yea, we only use Disconnect so far, if we need CoA later, we can add it, but for now Disconnect is all we need.

**Review by uri200 — APPROVED** [review]

**Comment by uri200** [comment]
lgtm

**Review by themarwhal — APPROVED** [review]
